### DISCUSSION TITLE
Airflow UI stops refreshing the DAG status

### DISCUSSION THREAD

**Discussion by pavelpi** [apache/airflow#34853] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Airflow versions 2.6.1, 2.6.2
When we run a DAG in Airflow, after a few seconds frontend stops updating DAg tasks statuses on UI and we cannot see DAG processing progress. When we watch communication between frontend and backend, we can see that suddenly the server started to report 404 on `grid_data?dag_id=<dag_id>&num_runs=25` with `{"error": "can't find dag <dag_id>"}`. It reports 404 twice and stops all further requests.
Refresh of the page causes page forward to DAGs landing page with error telling DAG <dag_id> seems to be missing from DagBag.
After a while (~1 minute) the DAG is discoverable again. In the background the DAG finishes normally.

### What you think should happen instead

The web frontend should keep reporting the DAG progress.

### How to reproduce

Run any long running DAG and watch its status.

### Operating System

python:3.10-slim-bullseye

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.0.0
apache-airflow-providers-apache-spark==4.0.1
apache-airflow-providers-celery==3.1.0
apache-airflow-providers-cncf-kubernetes==6.1.0
apache-airflow-providers-common-sql==1.4.0
apache-airflow-providers-elasticsearch==4.4.0
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-google==10.0.0
apache-airflow-providers-grpc==3.1.0
apache-airflow-providers-hashicorp==3.3.1
apache-airflow-providers-http==4.3.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-microsoft-azure==6.0.0
apache-airflow-providers-microsoft-mssql==3.3.2
apache-airflow-providers-mysql==5.0.0
apache-airflow-providers-odbc==3.2.1
apache-airflow-providers-opsgenie==5.0.0
apache-airflow-providers-oracle==3.6.0
apache-airflow-providers-postgres==5.4.0
apache-airflow-providers-redis==3.1.0
apache-airflow-providers-sftp==4.2.4
apache-airflow-providers-sqlite==3.3.2
apache-airflow-providers-ssh==3.6.0

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by jscheffl** [comment]
This feels like a side effect that if DAG parsing has performance problems, DAGs might dis-appear from the system until re-parsed. In case of Error the UI stops refreshing.

Can you please check your DAG parser performance and if DAGs dis- and re-appear?

**Comment by pavelpi** [comment]
Hello @jens-scheffler-bosch
thanks for quick reply!
I scaled up the environment to have 3 DAG processors, each one having 4 CPUs and 2GB memory. 70% of the time all CPUs are almost fully allocated. There are periods when I can see the same issue. It's better than when I had only 2 processors, 1 CPU each. I also tried 1 processor running on 8CPU/8GB RAM machine, but this did not help much. Memory does not seem to be a problem - processors need about 400MB each regardless of number of processors or #CPUs.
Is there any recommended setup or environment setting that would help us?
We have 137 DAGs, DAG processor processes them between 0.5s and 2.0s each.
When we encounter the issue, although Airflow reports missing DAG, it is still available in the list of DAGs.

**Comment by jscheffl** [comment]
Most important question: Problem gone by increasing DagProcessor resources? Otherwise this should be fixed first.

Besides throwing more hardware on the problem you can also configure other timeouts when DAGs are purged if not re-parsed. See configuration reference for options.
Otherwise if you see performance problems and DAG parsing is slow, would be good to take alook to best practices to improve DAG parsing performance. Typical root cause is Top level code (see https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code) and Variables which are adding delay because of DB round trips.

If you have taken a look to DAG processor logs and the times are really in the range of 0.5-2s per DAG then really something is slow in your DAGs. Either is it not good code or your DAGs are very complex. Good is if DAGs are parsed in <0.1 seconds... it is just the parsing of the DAG not the execution.

If okay, in general I'd propose to close the issue.

**Comment by Taragolis** [comment]
It would be nice if you could provide some DAG samples which might reproduce an issue, otherwise it is looks like more troubleshooting rather than actual issue.

**Comment by pavelpi** [comment]
The error occurs also if I remove all DAGs except for one run it then. In such case the dag processor is almost idle.

We are migrating from Airflow 1.10.15 where we have never seen such a problem.

Pasting DAG code here would not help much as most of our DAGs use our own common libraries and I cannot paste here the whole dependency chain. We are trying to be careful about toplevel code in our libraries.

I had assumption that Airflow frontend should be responsive even if periodic backend parsing would take longer time.

Is there any combination of configuration parameter we should pay attention to?

When the frontend stops responding we can see 2 consecutive failed calls against API. One isolated failure is recoverable.
![image](https://github.com/apache/airflow/assets/4568910/39009b98-ddc9-4f55-8275-69d0d0f36227)

![image](https://github.com/apache/airflow/assets/4568910/27c13104-9f94-4bc1-9e20-14408a5866d6)

**Reply by Taragolis** [reply]
>We are migrating from Airflow 1.10.15 where we have never seen such a problem.

DAGs which might work well in 1.10 might not work with the same performance/expectations in 2.x, especially if it use some kind of dynamic dag generation based on non-deterministic things.

In general I would recommend try to [fine tune your scheduler](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#fine-tuning-your-scheduler-performance) as well as try to increase [`[core] dagbag_import_timeout`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#dagbag-import-timeout)

But it could be also a different problem, you need to check your DagProcessor(s)/Scheduler logs

**Reply by pavelpi** [reply]
@Taragolis We are no using dynamic DAG generation. DAGs that fail are static ones.
I understand that there could be performance differences between Airflow versions, but there should not be periods when Airflow malfunctions.
As I stated below, it looks like web backend issue. Both scheduler(s) and worker(s) do not report any problems and run fine in the background.

**Reply by Taragolis** [reply]
> but there should not be periods when Airflow malfunctions.

The is no guarantee at all about any SLA because Airflow is OSS project and mostly developed by users of Airflow on their free time. Also many different factors could involve to you situation, which out of the Airflow control:
- Some hardware degradations or Cloud Resource throttling / limitations
- Some software issues, e.g. broken NTP sync
- Broken python dependancies
- Usage parts which doesn't cover by [Public Interface of Airflow](https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html)

This is not a complete list

>  it looks like web backend issue

Then you should look at webserver logs / metrics. Right now there is not enough information to figure out what is going wrong in your case.

**Reply by potiuk** [reply]
We have not enouh datapoints to help you with your problem @pavelpi (and this is likely your problem not Airflow because otherwise many people would have the same problem) .

I think the problem is very likely in the solution you implemented to synchronize the dags to dags volume. It's likely broken. Generally similar behaviour will occur if your synchronizing solution will delete and recreate the files - so that DAG file process will suddenly see empty or incomplete list of files in the DAG folder. I've seen people doing syncing in  the way that they removed all dags and copied new files to the new place. That led to dag file processing seeing - for split second - empty folder and deactivating the dags (marking them as deleted) - that would certainly lead to the behaviour you see. 

This could also be side effect of whatever syncing solution you use - some people used gcsfuse or s3 fuse or some other solutions that provide pretty unstable (network-bound) solutions. For example EFS without paying for iops will produce similar results - because "network filesystems" are tricky to  handle and you start getting network IOPS limitations on the EFS volumes, weird things happen.

Of course this is a wild guess, but I'd say - your syncing is broken (simiilarly to you wild guessing airflow webserver is). 

If I were you - I'd look there.

**Reply by pavelpi** [reply]
Thanks @taragolis and @potiuk for your comments.

* The DAG is not broken, because this would be shown on UI differently (error message with call stack). DAG is always being processed in the background without any issues.
* The problem happens in two our environments - AWS cloud and Github codespaces - exactly the same way. In first one we use standard git-sync, in the other (development environment) we use direct volume mapping. From this perspective we think the problem is not in file syncing.

I was monitoring database activity and found that these calls are being launched periodically by Airflow:
DELETE FROM serialized_dag
INSERT INTO serialized_dag (dag_id, fileloc, fileloc_hash, data, data_compressed, last_updated, dag_hash, processor_subdir) VALUES...

When I run the DAG and watch availability of it in serialized_dag database table, there are periods (~40 seconds) when this dag is not there. Exactly during this period the frontend always fails. See the image I posted previously regarding grid view and 404 errors. Similar can be achieved when I navigate to Graph view. Here the server responds with _500 Internal Server Error_ while the dag is not available in serialized_dag. Compared to Grid view the frontend is able to recover.

![image](https://github.com/apache/airflow/assets/4568910/4483c350-1173-4012-84c6-7aaf967b92d5)

This time I got error stack trace in webserver log: 

```
[2023-11-01T12:37:32.214+0000] {dagbag.py:211} WARNING - Serialized DAG connection_check no longer exists
[2023-11-01T12:37:32.215+0000] {app.py:1744} ERROR - Exception on /object/task_instances [GET]
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/www/auth.py", line 47, in decorated
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/www/decorators.py", line 125, in wrapper
    return f(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/www/views.py", line 3596, in task_instances
    for ti in dag.get_task_instances(dttm, dttm)
AttributeError: 'NoneType' object has no attribute 'get_task_instances'
10.240.4.209 - - [01/Nov/2023:12:37:32 +0000] "GET /object/task_instances?dag_id=connection_check&execution_date=2023-11-01T12%3A37%3A05.479619%2B00%3A00 HTTP/1.1" 500 1539 "https://humble-tribble-65pqp5rwxwfr4vx-31369.app.github.dev/dags/connection_check/graph?root=" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36"
```
because the dag received from `get_airflow_app().dag_bag.get_dag(dag_id)` returns None. It does not exist in dagbag at this time.

I went through similar published complaints, but could not find any solution that would work in our case.
One thing to point out is that the DAG tends to be unavailable only while it is running. The DAG is always available in serialized_dag when Airflow is idle.

**Reply by pavelpi** [reply]
When Airflow is in idle mode and I just watch finished DAG in Graph view, Something I receive 404

![image](https://github.com/apache/airflow/assets/4568910/5fa8d828-23ed-4f8d-9350-d7a3a23961bf)

and the frontend shortly shows 

![image](https://github.com/apache/airflow/assets/4568910/7e9037dd-ba75-4d39-b6a5-25afad2c35ef)

**Comment by pavelpi** [comment]
I wrote a script that periodically serialized 2 outputs
* API call to get the DAG
* `grid_data?dag_id=connection_check&num_runs=25`

In the first case the API always returned the DAG information, the second call had periods when it returned only `{"error": "can't find dag connection_check"}`
It looks like a bug in webserver backend

**Reply by potiuk** [reply]
Yes. Airflow scheduler or dag file processor (if standalone) will delete  dags fro serialized table precisely when it sees that dag files are missing OR when it fails to create the dags when parsing your DAG file. So what you likely see is instability of your dynamic DAG file - probably because of some networking issue or a bug - sometimes it produces DAg with an ID and sometimes it does not. Maybe network timeout that is handled by skipping dag creation, maybe a bug or race condition in your DAG code.you need to investigate and review your code. Every time the DAG file is parsed it should ALWAYS consistently produce the same set of dags with the same dag IDS. If it does not - you will get those kind of behaviours

**Comment by pavelpi** [comment]
@potiuk @Taragolis We upgraded to Airflow 2.7.2 and cannot reproduce the issue anymore. It was most likely fixed between 2.6.2 and 2.7.2

**Reply by potiuk** [reply, marked as answer]
Hmm. Thanks for confirming. There are many changes bettween these two and it's hard to say which one could have caused it but this only confirms that "upgrade to see if fixed" is a very viable way of dealing with such problems.
